lib/ansible/plugins/strategy/linear.py

```diff
@@ -355,25 +355,16 @@ def run(self, iterator, play_context):
                 failed_hosts = []
                 unreachable_hosts = []
                 for res in results:
-                    # execute_meta() does not set 'failed' in the TaskResult
-                    # so we skip checking it with the meta tasks and look just at the iterator
-                    if (res.is_failed() or res._task.action in C._ACTION_META) and iterator.is_failed(res._host):
+                    if res.is_failed():
                         failed_hosts.append(res._host.name)
                     elif res.is_unreachable():
                         unreachable_hosts.append(res._host.name)
 
-                # if any_errors_fatal and we had an error, mark all hosts as failed
-                if any_errors_fatal and (len(failed_hosts) > 0 or len(unreachable_hosts) > 0):
-                    dont_fail_states = frozenset([IteratingStates.RESCUE, IteratingStates.ALWAYS])
+                if any_errors_fatal and (failed_hosts or unreachable_hosts):
                     for host in hosts_left:
-                        (s, _) = iterator.get_next_task_for_host(host, peek=True)
-                        # the state may actually be in a child state, use the get_active_state()
-                        # method in the iterator to figure out the true active state
-                        s = iterator.get_active_state(s)
-                        if s.run_state not in dont_fail_states or \
-                           s.run_state == IteratingStates.RESCUE and s.fail_state & FailedStates.RESCUE != 0:
+                        if host.name not in failed_hosts:
                             self._tqm._failed_hosts[host.name] = True
-                            result |= self._tqm.RUN_FAILED_BREAK_PLAY
+                            iterator.mark_host_failed(host)
```

Comment:
we cannot do this before rescue on original failed host is attempted as it breaks the normal flow of hosts:

any_errors_fatal: true
- hostA fails => go to rescue if it exists, otherwise proceed as if rescue failed
- if rescue, hosts[b-n] should wait for rescue result (meta: noop filters?)
- hostA fails rescue/no rescue existed => fail hosts[b-n], always/nested rescue/always still need to be executed
- hostA succeeds rescue => hsots[b-n] proceed from hosta failed task as normal

Reply:
Were you able to create a play with `any_errors_fatal: true` that would not work as expected with this PR? I just tried and could not, it seems that `PlayIterator` handles the cases I tried correctly and rescues (or not) the hosts.